build.gradle

```diff
@@ -27,54 +25,55 @@ dependencies {
     compile("org.flywaydb:flyway-core:4.0.3")
 }
 
-/**
- * Docker test database configuration
- */
 import com.bmuschko.gradle.docker.tasks.container.*
 import com.bmuschko.gradle.docker.tasks.image.*
+
 task buildTestDatabaseImage(type: DockerBuildImage) {
     inputDir = file('docker')
+    remove true
     tag = 'test/database'
 }
+
 task createTestDatabaseContainer(type: DockerCreateContainer) {
     dependsOn buildTestDatabaseImage
     targetImageId { buildTestDatabaseImage.getImageId() }
     portBindings = ['5432:5432']
     containerId = dockerDatabaseId
     containerName = dockerDatabaseName
     doLast {
-        println '\nheyhey I\'m inside createTestDatabaseContainer\n'
+        println "I'm inside createTestDatabaseContainer"
     }
 }
-task startTestDatabaseContainer(type: DockerStartContainer) {
-    dependsOn createTestDatabaseContainer
+
+task removeTestDatabaseContainer(type: DockerRemoveContainer) {
     targetContainerId { dockerDatabaseId }
     doLast {
-        println '\nheyhey I\'m inside startTestDatabaseContainer\n'
+        println "I'm inside removeTestDatabaseContainer"
     }
 }
-startTestDatabaseContainer.doLast {
-    // Add some delay so that the database has time to start. TODO: Find a way to check that the database has started.
-    sleep 6000
-}
+
 task stopTestDatabaseContainer(type: DockerStopContainer) {
+    targetContainerId { dockerDatabaseId }
+    finalizedBy removeTestDatabaseContainer
     doLast {
-        println '\nheyhey I\'m inside stopTestDatabaseContainer\n'
+        println "I'm inside stopTestDatabaseContainer"
     }
-    targetContainerId { dockerDatabaseId }
 }
-task removeTestDatabaseContainer(type: DockerRemoveContainer) {
-    dependsOn stopTestDatabaseContainer
+
+task startTestDatabaseContainer(type: DockerStartContainer) {
+    dependsOn createTestDatabaseContainer
+    targetContainerId { dockerDatabaseId }
+    finalizedBy stopTestDatabaseContainer
```

Comment:
this will stop the container __just after__ it was stared and 6s elapsed...

Reply:
I'm using `startTestDatabaseContainer` for other things as well, so I'm sure if I can have a `finalizedBy stopTestDatabaseContainer` in `startTestDatabaseContainer`.

Reply:
@orzeh, even if you change 6s to 1s it runs perfectly fine. Maybe I just don't understand something but it works right now. If you can provide a command that you run that shows this PR fails it will be very useful.

Reply:
@Opalo you start DB container in order to do some Flyway migrations, so you can't stop that container just after starting it. It must run till migration ends. Then you can stop it. In Gradle tasks are executed one by one so when migration is done, `StartTestDatabaseContainer` task is already finished (finalized). There is a difference between finalizing task and stopping container and removing it. I hope you understand what I mean.

Reply:
@orzeh, yes I know tha of course. However, have you tried it? The container is started, the migrations are run and then container is stopped. I know how it should work - if it didn't play its role I wouldn't bother you at all with this PR. Once again - have you run it?

Reply:
Here is output from my terminal, is that how it should look like?
```
opal@opal-mac:~/t/g/flyway-migration-error-causes-final-gradle-task-to-not-execute(master|✚1) $ gradle validateMigration                                                                                                                                                   09:09
Step 1/5 : FROM postgres:9.5-alpine
---> 15492ded96c9
Step 2/5 : EXPOSE 5432
---> Using cache
---> 3229335f99f2
Step 3/5 : ENV POSTGRES_DB myDatabase
---> Using cache
---> 97ef34ec5550
Step 4/5 : ENV POSTGRES_USER myDatabaseUser
---> Using cache
---> a9ecccbe6ee2
Step 5/5 : ENV POSTGRES_PASSWORD myDatabasePassword
---> Using cache
---> 2fe07809d000
Successfully built 2fe07809d000
Successfully tagged test/database:latest

> Task :buildTestDatabaseImage
Building image using context '/Users/opal/tutorial/gradle/flyway-migration-error-causes-final-gradle-task-to-not-execute/docker'.
Using tag 'test/database' for image.
Created image with ID '2fe07809d000'.

> Task :createTestDatabaseContainer
Created container with ID '000ff1f3e89c713dcff44b6a86fe13a48a9656f01c1f43d9d938f616572eacb4'.
I'm inside createTestDatabaseContainer

> Task :startTestDatabaseContainer
Starting container with ID 'my_docker_test_database'.
I'm inside startTestDatabaseContainer

> Task :flywayMigrate
Gradle now uses separate output directories for each JVM language, but this build assumes a single directory for all classes from a source set. This behaviour has been deprecated and is scheduled to be removed in Gradle 5.0
Migration of schema "public" to version 1 - Create person table failed! Changes successfully rolled back.

> Task :stopTestDatabaseContainer
Stopping container with ID 'my_docker_test_database'.
I'm inside stopTestDatabaseContainer

> Task :removeTestDatabaseContainer
Removing container with ID 'my_docker_test_database'.
I'm inside removeTestDatabaseContainer


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':flywayMigrate'.
> Error occurred while executing flywayMigrate
  
  Migration V1__Create_person_table.sql failed
  --------------------------------------------
  SQL State  : 42601
  Error Code : 0
  Message    : ERROR: INSERT has more expressions than target columns
    Position: 58
  Location   : /Users/opal/tutorial/gradle/flyway-migration-error-causes-final-gradle-task-to-not-execute/src/main/resources/db/migration/V1__Create_person_table.sql (/Users/opal/tutorial/gradle/flyway-migration-error-causes-final-gradle-task-to-not-execute/src/main/resources/db/migration/V1__Create_person_table.sql)
  Line       : 6
  Statement  : INSERT INTO person (id, name) VALUES (1, 'Test Testson', 42)
  
  ERROR: INSERT has more expressions than target columns
    Position: 58

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED in 10s
6 actionable tasks: 6 executed
```

Reply:
That output indeed looks correct :)

Reply:
Ok, nevermind. Glad you've tried it out at all.

Reply:
Thanks again for the help Opalo!

Reply:
@Opalo I looks like I've learned something new about Gradle ;-)